### PR TITLE
Introduce boost qvm and a Quaternion type

### LINKED ISSUE #2289: Replace custom rotation and vector code by boost.qvm
[boost.qvm](https://www.boost.org/doc/libs/1_68_0/libs/qvm/doc/index.html) has most of the vector/matrix/quaternion methods that we need. With this we could drop most of our own quaternion and vector code, plus we would gain more things, like matrix-vector ops. Boost.qvm can work with out types by providing some simple template specializations, so this could be rolled out gradually and easily.

This would need to increase the required boost version to 1.62.0.

### PR DESCRIPTION
**Pull request by KaiSzuttor**
Fixes #2289

### CODE REVIEW

File: src/utils/include/utils/quaternion.hpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+#define SRC_UTILS_INCLUDE_UTILS_QUATERNION_HPP
+
+#include <boost/qvm/deduce_scalar.hpp>
+#include <boost/qvm/quat_operations.hpp>
+#include <boost/qvm/quat_traits.hpp>
+
+#include "utils/Vector.hpp"
+
+namespace Utils {
+template <typename T> class Quaternion : private Vector<T, 4> {
```
Comment (fweik) [inline src/utils/include/utils/quaternion.hpp]: Couldn't you just use the quaternion class from qvm?

Comment (KaiSzuttor) [reply src/utils/include/utils/quaternion.hpp]: you mean this?

``` c++
namespace boost { namespace qvm {

    template <class T>
    struct quat {

      T a[4];

      template <class R>
      operator R() const {
        R r;
        assign(r,*this);
        return r;
      }

    };
}
}

Comment (fweik) [reply src/utils/include/utils/quaternion.hpp]: Not enough `enable_if`?

File: src/utils/tests/quaternion_test.cpp [hunk 2]
```diff
@@ -1,7 +1,6 @@
 /*
  * Copyright (C) 2019 The ESPResSo project
  *
- * This file is part of ESPResSo.
```
Comment (jngrad) [inline src/utils/tests/quaternion_test.cpp]: accidental line deletion

File: src/utils/include/utils/quaternion.hpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
+#ifndef SRC_UTILS_INCLUDE_UTILS_QUATERNION_HPP
```
Comment (jngrad) [inline src/utils/include/utils/quaternion.hpp:19]: missing license header

**Comment by fweik** [comment]
I'd find it helpful if you would clarify what the final goal of this is. Do you want to do all vector/... operations via qvm? That type should come from where? E.g. it is not clear to my why you redfinde the free function operator overloads for `Quaternion` instead of just inlinening those of qvm, which you also use for the implementation. This looks odd to me, but I'm sure there is a good reason, albeit it not being spelled out. Also it has to be clear that what ever is done in this space will be very hard to change in the future...

**Comment by KaiSzuttor** [comment]
the reason is that this is a draft...

**Comment by fweik** [comment]
Yes, which is why it seems to me the right time to discuss the design. I don't think it helps towards a constructive design process that you are so defensive.

**Comment by KaiSzuttor** [comment]
i just use this draft PR for CI feedback, thanks for your feedback

File: src/core/Particle.hpp [hunk 4]
```diff
@@ -128,9 +129,10 @@ struct ParticleProperties {
     int to_particle_id = 0;
     double distance = 0;
     /** Relative position of the virtual site. */
-    Utils::Vector4d rel_orientation = {0., 0., 0., 0.};
+    Utils::Quaternion<double> rel_orientation =
+        Utils::Quaternion<double>::zero();
     /** Orientation of the virtual particle in the body fixed frame. */
-    Utils::Vector4d quat = {0., 0., 0., 0.};
+    Utils::Quaternion<double> quat = Utils::Quaternion<double>{0., 0., 0., 0.};
```
Comment (jngrad) [inline src/core/Particle.hpp]:
```suggestion
    Utils::Quaternion<double> quat = Utils::Quaternion<double>::zero();
```

File: src/utils/include/utils/quaternion.hpp [hunk 5]
```diff
@@ -0,0 +1,136 @@
+/*
```
Comment (jngrad) [inline src/utils/include/utils/quaternion.hpp:1]: should this file be renamed `Quaternion.hpp`?

**Review by jngrad — COMMENTED** [review]
besides minor changes, LGTM

File: src/utils/include/utils/quaternion.hpp (hunk 5 above)
Comment (KaiSzuttor) [reply src/utils/include/utils/quaternion.hpp:1]: i think we should not use capital letters in filenames at all... that lead to issues on case-insensitive file systems on windows or mac

**Review by jngrad — APPROVED** [review]

**Pull request merged**